rhine-terminal/src/FRP/Rhine/Terminal.hs

```diff
@@ -0,0 +1,165 @@
+{- | Wrapper to write @terminal@ applications in Rhine, using concurrency.
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeFamilies #-}
+module FRP.Rhine.Terminal
+  ( TerminalConcT
+  , TerminalEventClock (..)
+  , launchTerminal
+  , flowTerminal
+  , terminalConcurrently
+  )
+  where
+
+-- base
+import Prelude hiding (putChar)
+import Control.Concurrent ()
+import Control.Concurrent.MVar ()
+import Data.IORef ()
+
+-- time
+import Data.Time.Clock ( getCurrentTime )
+
+-- transformers
+import Control.Monad.Trans.Class ( MonadTrans )
+import Control.Monad.Trans.Reader ( ReaderT(..), ask )
+
+-- terminal
+-- terminal
+import System.Terminal
+import System.Terminal.Internal ( LocalTerminal )
+
+-- rhine
+import FRP.Rhine
+    ( MonadIO(..),
+      constM,
+      liftTransS,
+      flow,
+      concurrently,
+      Arrow(first),
+      Clock(..),
+      HoistClock(..),
+      GetClockProxy,
+      In,
+      Out,
+      Schedule(..),
+      Rhine,
+      UTCTime )
+import Control.Monad.Catch (MonadMask, MonadCatch, MonadThrow)
+
+-- | Wraps the LocalTerminal value needed for communication with the @terminal@ backend.
+newtype TerminalConcT m a = TerminalConcT
+  { unTerminalConcT :: ReaderT LocalTerminal m a }
+  deriving (Functor, Applicative, Monad, MonadTrans, MonadIO, MonadThrow, MonadCatch, MonadMask)
+
+instance MonadInput (TerminalConcT IO) where
+  awaitWith = awaitWith
+
+instance MonadPrinter (TerminalConcT IO) where
+  putLn = putLn
+  putChar = putChar
+  putText = putText
+  getLineWidth = getLineWidth
+
+instance MonadScreen (TerminalConcT IO) where
+  getWindowSize = getWindowSize
+  moveCursorUp = moveCursorUp
+  moveCursorDown = moveCursorDown
+  moveCursorForward = moveCursorForward
+  moveCursorBackward = moveCursorBackward
+  getCursorPosition = getCursorPosition
+  setCursorPosition = setCursorPosition
+  setCursorRow = setCursorRow
+  setCursorColumn = setCursorColumn
+  saveCursor = saveCursor
+  restoreCursor = restoreCursor
+  insertChars = insertChars
+  deleteChars = deleteChars
+  eraseChars = eraseChars
+  insertLines = insertLines
+  deleteLines = deleteLines
+  eraseInLine = eraseInLine
+  eraseInDisplay = eraseInDisplay
+  showCursor = showCursor
+  hideCursor = hideCursor
+  setAutoWrap = setAutoWrap
+  setAlternateScreenBuffer = setAlternateScreenBuffer
+
+-- * Terminal clock in 'IO'
+
+-- | Concurrently block on @gloss@ events.
+data TerminalEventClock = TerminalEventClock
```

Comment:
I'm thinking more and more that this is the right way. If you put the `LocalTerminal` in the clock, you could drop the `TerminalConcT` monad, make an `instance Clock IO TerminalEventClock`, and you don't need to write your own `Schedule` anymore. Yes, this would restrict the programs we can write now somewhat, but I think it's ok to start with until the dust has settled around other developments like https://github.com/turion/rhine/pull/148.

Reply:
Great! Happy to follow this path for now and follow new developments.